src/basic/pidref.c

```diff
@@ -8,6 +8,28 @@
 #include "pidref.h"
 #include "process-util.h"
 #include "signal-util.h"
+#include "stat-util.h"
+
+bool pidref_equal(const PidRef *a, const PidRef *b) {
+
+        if (pidref_is_set(a)) {
+                if (!pidref_is_set(b))
+                        return false;
+
+                if (a->pid != b->pid)
+                        return false;
+
+                /* pidfds live in their own pidfs and each process comes with a unique inode number since
+                 * kernel 6.8. We can safely do this on older kernels too though, as previously anonymous
+                 * inode was used and inode number was the same for all pidfds. */
+                if (a->fd >= 0 && b->fd >= 0 && fd_inode_same(a->fd, b->fd) <= 0)
```

Comment:
i think it would be safer to do `!= 0` rather than `<= 0` for the fd_inode_same() result check.

I am thinking here that fstatat() might fail for various reasons and we probably should trust the pid nr in that case. There recently was this hubbub for example where newer glibc version defaulted to use fstatat() internally even for the old stat()/fstat() library calls in environments where a seccomp filter blocked fstatat() however thus causing complete breakage. With that in the back of the head I think it's better to assume pids are OK for comparison if fstatat() can't work rather than assume they are never equal.

Reply:
(maybe a debug log if this fails would be good though, to make this easier to track down)